src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,353 @@
+                    if (re.Topology == null)
+                        await re.WaitForTopologyUpdate(re._firstTopologyUpdate);
+
+                    var shard = TcpConnection.ShardedContext.GetShardedDatabaseName(i);
+
+                    // we want to limit the batch of each shard, to not hold too much memory if there are other batches while batch is proceed
+                    if (_options.MaxDocsPerBatch == 4096) // this is default value
+                    {
+                        _options.MaxDocsPerBatch = 1024; //TODO: egor add configuration for shard worker MaxDocsPerBatch
+                        _options.MaxErroneousPeriod = TimeSpan.MaxValue; //TODO: egor we have reconnect with timeout now
+                    }
+
+                    var shardWorker = new ShardedSubscriptionWorker(_options, shard, re, SendConfirmToClientUnderLock, ReportExceptionToAndStopProcessing, WriteBatchToClientAndAck);
+
+                    var worker = new SubscriptionShardHolder
+                    {
+                        Worker = shardWorker,
+                        PullingTask = shardWorker.RunInternal(CancellationTokenSource.Token),
+                        RequestExecutor = re
+                    };
+
+                    _shardWorkers.Add(shard, worker);
+                }
+
+                var exceptions = new List<Exception>();
+                foreach (var worker in _shardWorkers)
+                {
+                    if (worker.Value.PullingTask.IsFaulted)
+                    {
+                        exceptions.Add(worker.Value.PullingTask.Exception);
+                        worker.Value.LastErrorDateTime = DateTime.Now;
+                    }
+                }
+
+                if (exceptions.Count == _shardWorkers.Count)
+                    throw new AggregateException($"Cannot start sharded subscription '{_options.SubscriptionName}' with id '{SubscriptionId}' for database '{TcpConnection.ShardedContext.DatabaseName}' because all shard workers failed.", exceptions);
+
+                await WriteJsonAsync(new DynamicJsonValue
+                {
+                    [nameof(SubscriptionConnectionServerMessage.Type)] = nameof(SubscriptionConnectionServerMessage.MessageType.ConnectionStatus),
+                    [nameof(SubscriptionConnectionServerMessage.Status)] = nameof(SubscriptionConnectionServerMessage.ConnectionStatus.Accepted)
+                }, TcpConnection);
+            }
+            catch
+            {
+                DisposeOnDisconnect.Dispose();
+                throw;
+            }
+        }
+
+        private async Task MaintainConnectionWithClientWorker()
+        {
+            using (DisposeOnDisconnect)
+            {
+                while (CancellationTokenSource.IsCancellationRequested == false)
+                {
+                    await TimeoutManager.WaitFor(TimeSpan.FromMilliseconds(3000), CancellationTokenSource.Token);
```

Comment:
Thinking about it, you are _unconditionally_ waiting for 3 seconds, whatever for?